casper/src/main/resources/PoS.rhox

```diff
@@ -86,24 +89,48 @@ new  PoS,
             } |
 
             contract PoS (@"pay", @amount, returnCh) = {
-              new vaultCh, transferCh, userCh, processCh, depositCh in {
+              new vaultCh, transferCh, userCh, processPayCh,
+                  depositCh,
+                  computeSum, totalStakeCh,
+                  computeDelta, rewardsDeltaCh,
+                  computeMergeMap, newRewardsCh in {
                 getUser!(*userCh) |
 
-                runMVar!(*stateCh, *processCh, *returnCh) |
+                runMVar!(*stateCh, *processPayCh, *returnCh) |
 
-                for(@(amounts, bonds), resultCh <- processCh;
+                for(@(rewards, bonds), payResultCh <- processPayCh;
                     @userPk <- userCh) {
                   deposit!(userPk, amount, *depositCh) |
+                  @ListOps!("fold", bonds.toList(), 0, *computeSum, *totalStakeCh) |
+                  contract computeSum(@(pk, (stake, _)), @acc, resultCh) = {
+                    resultCh!(acc + stake)
+                  } |
+
+                  for(@totalStake <- totalStakeCh) {
+                    @ListOps!("fold", bonds.toList(), {}, *computeDelta, *rewardsDeltaCh) |
+                    contract computeDelta(@(pk, (currentValidatorStake, _)), @acc, resultCh) = {
+                      resultCh!(acc.set(pk, (amount * currentValidatorStake) / totalStake))
+                    }
+                  } |
+
+                  for (@rewardsDelta <- rewardsDeltaCh) {
+                    @ListOps!("fold", rewardsDelta.toList(), rewards, *computeMergeMap, *newRewardsCh) |
+                    contract computeMergeMap(@(pk, rewardDelta), @newRewards, resultCh) = {
+                      resultCh!(newRewards.set(pk, newRewards.getOrElse(pk, 0) + rewardDelta))
```

Comment:
This `getOrElse` is used because, if it is the first deploy after a validator has been bonded, the validator has never before received a reward so the validator will not show up in the rewards map. Rather than handle the corner-case at the time the map is updated, you should just add each validator to the rewards map in the `bond` method. The logic is more clear that way: "when a validator is bonded, they are added to the bonds map and to the rewards map".